File: src/main/java/lekkit/scev/machine/MachineBackend.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine;

import org.jetbrains.annotations.Nullable;

/**
 * Abstraction over a virtual-machine implementation.
 *
 * <p>In production, this is backed by RVVM via JNI
 * ({@link lekkit.scev.machine.rvvm.RvvmMachineBackend}). In tests we use
 * {@link lekkit.scev.machine.test.FakeMachineBackend} to exercise every
 * lifecycle path without loading librvvm.
 *
 * <p>Lifecycle contract:
 * <pre>
 *   new backend -> initialize(spec) -> start() -> (pause()/reset() allowed) -> close()
 * </pre>
 * All methods return {@code false} if the backend isn't in the right state
 * (not yet initialized, already closed, etc.).
 *
 * <p>Attach-order and peripheral wiring are the backend's responsibility. The
 * caller gives a spec, the backend constructs everything the spec describes,
 * loads firmware if present, and returns success / failure. After initialize
 * succeeds, {@link #framebuffer()}, {@link #keyboard()}, {@link #mouse()}, and
 * {@link #gpio()} return non-null iff the spec requested that device.
 */
public interface MachineBackend extends AutoCloseable {
    /**
     * Construct the machine and attach all peripherals described by {@code spec}.
     * Must be called exactly once, before {@link #start()}. Returns {@code true}
     * on success.
     */
    boolean initialize(MachineSpec spec);

    /** Start (or resume) execution. Returns {@code true} on success. */
    boolean start();

    /**
     * Halt the emulation thread. Does not "power off" the machine — subsequent
     * {@link #start()} resumes from the same point. {@link #isRunning()} stays
     * {@code true} across pause; only {@link #close()} flips it.
     */
    boolean pause();

    /** Reset the machine's CPU to its entry point. The caller must re-{@link #start()}. */
    boolean reset();

    /**
     * Machine is logically powered on — {@link #start()} has been called and
     * {@link #close()} has not. This mirrors RVVM's {@code machine_powered}
     * concept. A paused machine is still running in this sense.
     */
    boolean isRunning();

    /** Machine is in a valid initialized state (not null, not freed). */
    boolean isValid();

    /** The spec used to build this machine. */
    MachineSpec spec();

    /** The display, or null if this machine has no display. */
    @Nullable FramebufferView framebuffer();

    /** The keyboard, or null if keyboard wasn't created. */
    @Nullable KeyboardDevice keyboard();

    /** The mouse, or null if mouse wasn't created. */
    @Nullable MouseDevice mouse();

    /** The GPIO device, or null if no GPIO card was installed. */
    @Nullable GpioDevice gpio();

    /**
     * Direct read/write view into machine physical memory. Returns a mutable
     * {@link java.nio.ByteBuffer} whose backing storage IS the VM's memory —
     * writes are seen by the CPU, reads reflect CPU-written values.
     *
     * <p>Returns {@code null} if the backend doesn't support DMA (purely
     * simulated backends may return null) or the address range isn't mapped.
     *
     * <p>Used by:
     * <ul>
     *   <li>The bootrom installer to place a tiny demo program at the reset
     *       vector so the CPU has something to execute.</li>
     *   <li>E2E tests that need to verify the CPU actually ran code by
     *       reading back a memory-based side effect.</li>
     *   <li>Future firmware loaders that need to place initial memory state.</li>
     * </ul>
     */
    @Nullable java.nio.ByteBuffer readMemory(long addr, long size);

    /**
     * Tear down the machine and all attached peripherals. After close the
     * backend is unusable; attempting to call anything else returns false /
     * null. Safe to call multiple times.
     */
    @Override void close();
}

File: src/main/kotlin/lekkit/scev/machine/MachineBackend.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.machine

import java.nio.ByteBuffer

/**
 * Abstraction over a virtual-machine implementation.
 *
 * In production, this is backed by RVVM via JNI
 * ([lekkit.scev.machine.rvvm.RvvmMachineBackend]). In tests we use
 * `lekkit.scev.machine.test.FakeMachineBackend` to exercise every
 * lifecycle path without loading librvvm.
 *
 * Lifecycle contract:
 *
 * ```
 *   new backend -> initialize(spec) -> start() -> (pause()/reset() allowed) -> close()
 * ```
 *
 * All methods return `false` if the backend isn't in the right state
 * (not yet initialized, already closed, etc.).
 *
 * Attach-order and peripheral wiring are the backend's responsibility.
 * The caller gives a spec, the backend constructs everything the spec
 * describes, loads firmware if present, and returns success / failure.
 * After initialize succeeds, [framebuffer], [keyboard], [mouse], and
 * [gpio] return non-null iff the spec requested that device.
 */
interface MachineBackend : AutoCloseable {
    /**
     * Construct the machine and attach all peripherals described by
     * [spec]. Must be called exactly once, before [start]. Returns
     * `true` on success.
     */
    fun initialize(spec: MachineSpec): Boolean

    /** Start (or resume) execution. Returns `true` on success. */
    fun start(): Boolean

    /**
     * Halt the emulation thread. Does not "power off" the machine —
     * subsequent [start] resumes from the same point. [isRunning]
     * stays `true` across pause; only [close] flips it.
     */
    fun pause(): Boolean

    /** Reset the machine's CPU to its entry point. The caller must re-[start]. */
    fun reset(): Boolean

    /**
     * Machine is logically powered on — [start] has been called and
     * [close] has not. This mirrors RVVM's `machine_powered` concept.
     * A paused machine is still running in this sense.
     */
    fun isRunning(): Boolean

    /** Machine is in a valid initialized state (not null, not freed). */
    fun isValid(): Boolean

    /** The spec used to build this machine. */
    fun spec(): MachineSpec

    /** The display, or null if this machine has no display. */
    fun framebuffer(): FramebufferView?

    /** The keyboard, or null if keyboard wasn't created. */
    fun keyboard(): KeyboardDevice?

    /** The mouse, or null if mouse wasn't created. */
    fun mouse(): MouseDevice?

    /** The GPIO device, or null if no GPIO card was installed. */
    fun gpio(): GpioDevice?

    /**
     * The RPC serial device — a dedicated NS16550A whose chardev is a
     * JVM-visible ring buffer, exposed to the guest as a second UART
     * (`/dev/ttyS1` once the kernel console claims ttyS0). Used by
     * [lekkit.scev.rpc.ScevRpcManager] to carry COBS-framed MessagePack
     * RPC traffic.
     *
     * Returns `null` if the backend is purely simulated (test fakes)
     * or if the JNI bridge failed to attach.
     */
    fun serial(): SerialDevice?

    /**
     * Direct read/write view into machine physical memory. Returns a
     * mutable [ByteBuffer] whose backing storage IS the VM's memory —
     * writes are seen by the CPU, reads reflect CPU-written values.
     *
     * Returns `null` if the backend doesn't support DMA (purely
     * simulated backends may return null) or the address range isn't
     * mapped.
     *
     * Used by:
     * - The bootrom installer to place a tiny demo program at the reset
     *   vector so the CPU has something to execute.
     * - E2E tests that need to verify the CPU actually ran code by
     *   reading back a memory-based side effect.
     * - Future firmware loaders that need to place initial memory state.
     */
    fun readMemory(addr: Long, size: Long): ByteBuffer?

    /**
     * Tear down the machine and all attached peripherals. After close
     * the backend is unusable; attempting to call anything else
     * returns false / null. Safe to call multiple times.
     */
    override fun close()
}
